src/sentry/api/bases/monitor.py

```diff
@@ -57,23 +60,64 @@ class MonitorEndpoint(Endpoint):
     def respond_invalid() -> Response:
         return Response(status=status.HTTP_400_BAD_REQUEST, data={"details": "Invalid monitor"})
 
-    def convert_args(self, request: Request, monitor_id: str, *args, **kwargs):
-        try:
-            UUID(monitor_id)
-        except ValueError:
-            raise ParameterValidationError("Invalid monitor UUID")
+    def convert_args(
+        self,
+        request: Request,
+        monitor_id: str,
+        organization_slug: str | None = None,
+        *args,
+        **kwargs,
+    ):
+        organization = None
+        monitor = None
+
+        # The only monitor endpoints that do not have the org slug in their
+        # parameters are the GUID-style checkin endpoints
+        if organization_slug:
+            try:
+                organization = Organization.objects.get_from_cache(slug=organization_slug)
+            except Organization.DoesNotExist:
+                pass
+
+        # Try lookup by slug first. This requires organization context since
+        # slugs are uniuqe only to the organization
+        if organization:
+            try:
+                monitor = Monitor.objects.get(organization_id=organization.id, slug=monitor_id)
+            except Monitor.DoesNotExist:
+                pass
 
-        try:
-            monitor = Monitor.objects.get(guid=monitor_id)
-        except Monitor.DoesNotExist:
+        # Try lookup by GUID
+        if not monitor:
+            # Validate GUIDs
+            try:
+                UUID(monitor_id)
+            except ValueError:
+                # This error is a bit confusing, because this may also mean
+                # that we've failed to lookup their monitor by slug.
+                raise ParameterValidationError("Invalid monitor UUID")
+            # When looking up by guid we don't include the org conditional
+            # (since GUID lookup allows orgless routes), we will validate
+            # permissions later in this function
+            try:
+                monitor = Monitor.objects.get(guid=monitor_id)
+            except Monitor.DoesNotExist:
+                monitor = None
+
+        if not monitor:
             raise ResourceDoesNotExist
 
         project = Project.objects.get_from_cache(id=monitor.project_id)
         if project.status != ProjectStatus.VISIBLE:
             raise ResourceDoesNotExist
 
         if hasattr(request.auth, "project_id") and project.id != request.auth.project_id:
-            raise InvalidAuthProject
+            raise ResourceDoesNotExist
+
+        # When looking up via GUID we do not check the organiation slug,
+        # validate that the slug matches the org of the monitors projet
+        if organization_slug and project.organization.slug != organization_slug:
+            raise ResourceDoesNotExist
```

Comment:
if we're only going to check this for GUIDs, this should probably live under that GUID check logic

Reply:
I don't really mind this here since we need the `project` anyway.

I am trying to keep this logic not TOO complicated since there's a 2x2 matrix of things (guid vs slug, org slug vs no org slug)

Reply:
ah I guess I didn't realize about org_slug with guid